### PR TITLE
De-hardcode cable tiers (again).

### PR DESCRIPTION
**Pull request by Fuyukai**
Updated onto the 1.20 branch. ~~Still not complete, but~~ takes into account some of the comments from last time.

- Newly added tiers sort themselves automatically, with transformers between new and old tiers Just Working.
- Newly added tier blocks... will crash the game due to missing models. Runtime datagen doesn't appear to work properly in this version, either, so...

### CODE REVIEW

File: src/main/java/aztech/modern_industrialization/api/energy/CableTier.java [hunk 1]
```diff
@@ -25,43 +25,55 @@
...
 import aztech.modern_industrialization.MIText;
 import net.minecraft.network.chat.Component;
-import org.jetbrains.annotations.ApiStatus;
+import net.minecraft.network.chat.MutableComponent;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
```
Comment (James103) [inline src/main/java/aztech/modern_industrialization/api/energy/CableTier.java:34]: Please re-order the imports as follows:
```java
import aztech.modern_industrialization.MIText;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.MutableComponent;
import org.jetbrains.annotations.NotNull;
```

File: src/main/java/aztech/modern_industrialization/api/energy/CableTier.java [hunk 2]
```diff
@@ -25,43 +25,55 @@
...
-    EV("EV", "ev", 32 * 4 * 8 * 8, MIText.CableTierEV),
-    SUPERCONDUCTOR("Superconductor", "superconductor", 128000000, MIText.CableTierSuperconductor);
+public final class CableTier implements Comparable<CableTier> {
+    // static references
+    public static CableTier LV = new CableTier("LV", "lv", 32, MIText.CableTierLV);
+    public static CableTier MV = new CableTier("MV", "mv", 32 * 4, MIText.CableTierMV);
+    public static CableTier HV = new CableTier("HV", "hv", 32 * 4 * 8, MIText.CableTierHV);
+    public static CableTier EV = new CableTier("EV", "ev", 32 * 4 * 8 * 8, MIText.CableTierEV);
+    public static CableTier SUPERCONDUCTOR = new CableTier("Superconductor", "superconductor", 128000000, MIText.CableTierSuperconductor);
+
```
Comment (James103) [inline src/main/java/aztech/modern_industrialization/api/energy/CableTier.java:46]:
```suggestion
```

File: src/main/java/aztech/modern_industrialization/api/energy/CableTier.java [hunk 3]
```diff
@@ -25,43 +25,55 @@
...
+    }
+
+    /**
+     * @return A copy of the english name text component stored within.
+     */
+    public MutableComponent getEnglishNameComponent() {
+        return englishNameComponent.copy();
     }
 
+
```
Comment (James103) [inline src/main/java/aztech/modern_industrialization/api/energy/CableTier.java:76]:
```suggestion
```

File: src/main/java/aztech/modern_industrialization/api/energy/CableTier.java [hunk 4]
```diff
@@ -74,17 +86,67 @@ public long getEu() {
...
+        TIERS.put(tier.name, tier);
+    }
+
+    /**
+     * Looks up a cable tier registered by name.
+     *
+     * @param name The name of the tier, e.g. <code>LV</code>.
+     * @return An instance of the cable tier with the provided name.
+     */
+    @NotNull  // Should it be?
```
Comment (James103) [inline src/main/java/aztech/modern_industrialization/api/energy/CableTier.java:122]:
```suggestion
    @NotNull // Should it be?
```

File: src/main/java/aztech/modern_industrialization/machines/init/SingleBlockSpecialMachines.java [hunk 5]
```diff
@@ -31,6 +31,8 @@
 import aztech.modern_industrialization.machines.models.MachineCasing;
 import aztech.modern_industrialization.machines.models.MachineCasings;
```
Comment (James103) [inline src/main/java/aztech/modern_industrialization/machines/init/SingleBlockSpecialMachines.java:33]:
```suggestion
```

**Review by James103 — COMMENTED** [review]
The build is failing due to some errors in the formatting of the files.
Please run `./gradlew :spotlessApply` on the code or fix the errors manually, then this PR can compile.

**Comment by Fuyukai** [comment]
Yes, I am aware. I will fix it when I am done.

File: src/main/java/aztech/modern_industrialization/compat/kubejs/material/MIMaterialKubeJSEvents.java [hunk 6]
```diff
@@ -32,4 +32,5 @@ public interface MIMaterialKubeJSEvents {
 
     EventHandler ADD_MATERIALS = EVENT_GROUP.startup("addMaterials", () -> AddMaterialsEventJS.class);
     EventHandler MODIFY_MATERIAL = EVENT_GROUP.startup("modifyMaterial", () -> ModifyMaterialEventJS.class).extra(Extra.STRING);
+    EventHandler ADD_CABLE_TIERS = EVENT_GROUP.startup("cableTiers", () -> AddCableTiersEventJS.class);
```
Comment (Fuyukai) [inline src/main/java/aztech/modern_industrialization/compat/kubejs/material/MIMaterialKubeJSEvents.java:35]: Wasn't too sure where to put this; went with materials (as its mostly related to that). Definitely open to changing it if anyone has a better idea.

**Comment by Fuyukai** [comment]
There was a comment on the last PR about changing the EU values for the built-in tiers as well; I didn't add that here but I can make a second PR after this one is merged to add support for that.

File: src/main/java/aztech/modern_industrialization/compat/kubejs/material/MIMaterialKubeJSEvents.java (hunk 6 above)
Comment (Technici4n) [reply src/main/java/aztech/modern_industrialization/compat/kubejs/material/MIMaterialKubeJSEvents.java]: I moved it to registration events

**Comment by Technici4n** [comment]
Sorry, I didn't mean to leave this PR like that for 4 months. Thanks for providing the initial prototype. I tweaked the PR a bit and will merge it now.

**Pull request merged**